src/stochastic_swap.rs

```diff
@@ -0,0 +1,257 @@
+            cost_reduced = false;
+            // Loop over edges of coupling graph
+            need_copy = true;
+            for idx in 0..num_edges {
+                let start_edge = edges[2 * idx];
+                let end_edge = edges[2 * idx + 1];
+                let start_qubit = trial_layout.phys_to_logic[start_edge];
+                let end_qubit = trial_layout.phys_to_logic[end_edge];
+                if qubit_set.contains(&start_qubit) && qubit_set.contains(&end_qubit) {
+                    // Try this edge to reduce cost
+                    if need_copy {
+                        new_layout = trial_layout.clone();
+                        need_copy = false;
+                    }
+                    new_layout.swap(start_edge, end_edge);
+                    // compute objective function
+                    new_cost = compute_cost(&scale.view(), &new_layout, gates, num_gates);
+                    // record progress if we succeed
+                    if new_cost < min_cost {
+                        cost_reduced = true;
+                        min_cost = new_cost;
+                        optimal_layout = new_layout.clone();
+                        optimal_start = start_edge;
+                        optimal_end = end_edge;
+                        optimal_start_qubit = start_qubit;
+                        optimal_end_qubit = end_qubit;
+                        need_copy = true;
+                    } else {
+                        new_layout.swap(start_edge, end_edge);
+                    }
+                }
+            }
+            // After going over all edges
+            // Were there any good swap choices?
+            if cost_reduced {
+                qubit_set.remove(&optimal_start_qubit);
+                qubit_set.remove(&optimal_end_qubit);
+                trial_layout = optimal_layout.clone();
+                opt_edges.add(optimal_start, optimal_end);
+            } else {
+                break;
+            }
+        }
+        // We have either run out of swap pairs to try or failed to improve
+        // the cost
+
+        // Compute the coupling graph distance
+        dist = compute_cost(&cdist, &trial_layout, gates, num_gates);
+        // If all gates can be applied now we're finished.
+        // Otherwise we need to consider a deeper swap circuit
+        if dist as usize == num_gates {
+            break;
+        }
+        // increment the depth
+        depth_step += 1;
+    }
+    // Either we have succeeded at some depth d < d_max or failed
+    dist = compute_cost(&cdist, &trial_layout, gates, num_gates);
+    if dist as usize == num_gates && depth_step == 1 {
+        let mut best_possible = locked_best_possible.write().unwrap();
+        **best_possible = Some((dist, opt_edges, trial_layout, depth_step));
+        return None;
+    }
+    Some((dist, opt_edges, trial_layout, depth_step))
+}
+
+#[pyfunction]
+#[pyo3(text_signature = "(graph, weight_fn, /)")]
```

Comment:
The `/` is only allowed in Python signatures since 3.8 - that won't cause us any issues here, right, because it's compiled in rather than Python-interpreted?

Reply:
I don't think it should be a problem as this just gets set as the `__text_signature__` attribute on the function object that gets generated by the macro. That being said I've never been able to find a clear answer on how it's used vs `__signature__` or explicitly how it gets parsed by python or sphinx. I do know setting it is necessary to get the signature to show up correctly in sphinx though.

As you can tell I just copied this from retworkx and at least over there I haven't heard any complaints or issues from py3.6 or py3.7 users around the signatures.